src/libstd/lazy.rs

```diff
@@ -0,0 +1,746 @@
+    }
+}
+
+#[unstable(feature = "once_cell", issue = "68198")]
+impl<T: Clone> Clone for Once<T> {
+    fn clone(&self) -> Once<T> {
+        let res = Once::new();
+        if let Some(value) = self.get() {
+            match res.set(value.clone()) {
+                Ok(()) => (),
+                Err(_) => unreachable!(),
+            }
+        }
+        res
+    }
+}
+
+#[unstable(feature = "once_cell", issue = "68198")]
+impl<T> From<T> for Once<T> {
+    fn from(value: T) -> Self {
+        let cell = Self::new();
+        cell.get_or_init(|| value);
+        cell
+    }
+}
+
+#[unstable(feature = "once_cell", issue = "68198")]
+impl<T: PartialEq> PartialEq for Once<T> {
+    fn eq(&self, other: &Once<T>) -> bool {
+        self.get() == other.get()
+    }
+}
+
+#[unstable(feature = "once_cell", issue = "68198")]
+impl<T: Eq> Eq for Once<T> {}
+
+impl<T> Once<T> {
+    /// Creates a new empty cell.
+    #[unstable(feature = "once_cell", issue = "68198")]
+    pub const fn new() -> Once<T> {
+        Once {
+            state_and_queue: AtomicUsize::new(INCOMPLETE),
+            _marker: PhantomData,
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+        }
+    }
+
+    /// Gets the reference to the underlying value.
+    ///
+    /// Returns `None` if the cell is empty, or being initialized. This
+    /// method never blocks.
+    #[unstable(feature = "once_cell", issue = "68198")]
+    pub fn get(&self) -> Option<&T> {
+        if self.is_initialized() {
+            // Safe b/c checked is_initialize
+            Some(unsafe { self.get_unchecked() })
+        } else {
+            None
+        }
+    }
+
+    /// Gets the mutable reference to the underlying value.
+    ///
+    /// Returns `None` if the cell is empty.
+    #[unstable(feature = "once_cell", issue = "68198")]
+    pub fn get_mut(&mut self) -> Option<&mut T> {
+        if self.is_initialized() {
+            // Safe b/c checked is_initialize and we have a unique access
+            Some(unsafe { self.get_unchecked_mut() })
+        } else {
+            None
+        }
+    }
+
+    /// Get the reference to the underlying value, without checking if the
+    /// cell is initialized.
+    ///
+    /// Safety:
+    ///
+    /// Caller must ensure that the cell is in initialized state, and that
+    /// the contents are acquired by (synchronized to) this thread.
+    #[unstable(feature = "once_cell", issue = "68198")]
+    pub unsafe fn get_unchecked(&self) -> &T {
+        debug_assert!(self.is_initialized());
+        (&*self.value.get()).get_ref()
+    }
+
+    /// Get the reference to the underlying value, without checking if the
+    /// cell is initialized.
+    ///
+    /// Safety:
+    ///
+    /// Caller must ensure that the cell is in initialized state, and that
+    /// the contents are acquired by (synchronized to) this thread.
+    #[unstable(feature = "once_cell", issue = "68198")]
+    pub unsafe fn get_unchecked_mut(&mut self) -> &mut T {
```

Comment:
This is a new method, and I don't feel very good about it. Something like `as_ptr` can be an alternative, like the other interior mutability types.

Reply:
I should probably word it a bit better (although you probably know)...

The basic promise that makes `OnceCell` safe is that is either one of these states:
- uninitalized, where the only valid action is to initialize it.
- initialized, where it is never modified again.

In both states handing out a mutable reference is invalid.

I do wonder about `get_unchecked`. It is useful if you know for sure the value is initialized, and really need the performance. If there comes an `as_ptr` method similar to `Cell`, `RefCell` and friends, it would be unnecessary. And only one method for unsafe use seems like enough?

Reply:
I think, for std, we should just remove `get_unchecked`, as its utility is marginal at best. I am more comfortable with having that method in `once_cell` crate, as it's easier to evolve their.

Reply:
I’ll remove both these methods :+1: But an `as_ptr` seems like a reasonable addition.